src/kvstore/comm.h

```diff
@@ -621,7 +659,63 @@ class CommDevice : public Comm {
                           const std::vector<std::pair<NDArray*, NDArray>>& dst,
                           const bool use_copy,
                           const int priority) override {
-    LOG(FATAL) << "Not implemented yet";
+    CHECK_EQ(src.storage_type(), kRowSparseStorage)
+      << "BroadcastRowSparse expects row-sparse src NDArray";
+
+    bool is_same_rowid = true;
+    for (size_t i = 1; i < dst.size(); ++i) {
+      if (dst[i].second.var() != dst[0].second.var()) {
+        is_same_rowid = false;
+      }
+    }
+
+    for (size_t i = 0; i < dst.size(); ++i) {
+      if (is_same_rowid && i != 0) {
+        CopyFromTo(*dst[0].first, dst[i].first, priority);
+        continue;
+      }
+
+      NDArray* out = dst[i].first;
+      NDArray row_id = dst[i].second;
+      if (use_copy) {
+        CopyFromTo(src, out, priority);
+      } else {
+        CHECK_EQ(out->storage_type(), kRowSparseStorage)
+                 << "BroadcastRowSparse expects row_sparse dst NDArray";
+        const bool is_diff_ctx = out->ctx() != src.ctx();
```

Comment:
Are we assuming src is always on GPU?
If so, should we perform retain first before copying it to other devices?

Reply:
src is not assumed to be on gpu. Actually src is always on cpu. As you can see in  https://github.com/apache/incubator-mxnet/blob/master/src/kvstore/kvstore_local.h#L233, src is `local_[key]`. And `local_[key]` is initialized to be on `pinned_ctx_` which is always cpu, https://github.com/apache/incubator-mxnet/blob/master/src/kvstore/kvstore_local.h#L152.

Reply:
That's true at the beginning. But as soon as you push some gradients on GPU, it copies the weight from pinned_ctx to GPU. See 
https://github.com/apache/incubator-mxnet/blob/master/src/kvstore/kvstore_local.h#L173

Reply:
Nonetheless, I think performing sparse retain before the copy makes more sense since the source array is usually very large.